bundles/org.openhab.binding.insteon/src/main/resources/OH-INF/thing/legacy-thing-types.xml

```diff
@@ -103,443 +108,444 @@
+	<channel-type id="legacyFanMode">
 		<item-type>Number</item-type>
 		<label>Fan Mode</label>
 	</channel-type>
 
-	<channel-type id="fastOnOff">
+	<channel-type id="legacyFastOnOff">
 		<item-type>Switch</item-type>
 		<label>Fast On/Off</label>
 	</channel-type>
 
-	<channel-type id="fastOnOffButtonA">
+	<channel-type id="legacyFastOnOffButtonA">
 		<item-type>Switch</item-type>
 		<label>Fast On/Off Button A</label>
-		<config-description-ref uri="channel-type:insteon:keypad-button-fastonoff"/>
+		<config-description-ref uri="channel-type:insteon:legacy-keypad-button-fastonoff"/>
 	</channel-type>
 
-	<channel-type id="fastOnOffButtonB">
+	<channel-type id="legacyFastOnOffButtonB">
 		<item-type>Switch</item-type>
 		<label>Fast On/Off Button B</label>
-		<config-description-ref uri="channel-type:insteon:keypad-button-fastonoff"/>
+		<config-description-ref uri="channel-type:insteon:legacy-keypad-button-fastonoff"/>
 	</channel-type>
 
-	<channel-type id="fastOnOffButtonC">
+	<channel-type id="legacyFastOnOffButtonC">
 		<item-type>Switch</item-type>
 		<label>Fast On/Off Button C</label>
-		<config-description-ref uri="channel-type:insteon:keypad-button-fastonoff"/>
+		<config-description-ref uri="channel-type:insteon:legacy-keypad-button-fastonoff"/>
 	</channel-type>
 
-	<channel-type id="fastOnOffButtonD">
+	<channel-type id="legacyFastOnOffButtonD">
 		<item-type>Switch</item-type>
 		<label>Fast On/Off Button D</label>
-		<config-description-ref uri="channel-type:insteon:keypad-button-fastonoff"/>
+		<config-description-ref uri="channel-type:insteon:legacy-keypad-button-fastonoff"/>
 	</channel-type>
 
-	<channel-type id="fastOnOffButtonE">
+	<channel-type id="legacyFastOnOffButtonE">
 		<item-type>Switch</item-type>
 		<label>Fast On/Off Button E</label>
-		<config-description-ref uri="channel-type:insteon:keypad-button-fastonoff"/>
+		<config-description-ref uri="channel-type:insteon:legacy-keypad-button-fastonoff"/>
 	</channel-type>
 
-	<channel-type id="fastOnOffButtonF">
+	<channel-type id="legacyFastOnOffButtonF">
 		<item-type>Switch</item-type>
 		<label>Fast On/Off Button F</label>
-		<config-description-ref uri="channel-type:insteon:keypad-button-fastonoff"/>
+		<config-description-ref uri="channel-type:insteon:legacy-keypad-button-fastonoff"/>
 	</channel-type>
 
-	<channel-type id="fastOnOffButtonG">
+	<channel-type id="legacyFastOnOffButtonG">
 		<item-type>Switch</item-type>
 		<label>Fast On/Off Button G</label>
-		<config-description-ref uri="channel-type:insteon:keypad-button-fastonoff"/>
+		<config-description-ref uri="channel-type:insteon:legacy-keypad-button-fastonoff"/>
 	</channel-type>
 
-	<channel-type id="fastOnOffButtonH">
+	<channel-type id="legacyFastOnOffButtonH">
 		<item-type>Switch</item-type>
 		<label>Fast On/Off Button H</label>
-		<config-description-ref uri="channel-type:insteon:keypad-button-fastonoff"/>
+		<config-description-ref uri="channel-type:insteon:legacy-keypad-button-fastonoff"/>
```

Comment:
This is why I added a way to store the current channel configuration in `InsteonLegacyNetworkHandler` which are restored during the `InsteonLegacyDeviceHandler` initialization, part of the migration process.

Newly created channels will use the new config description uri along with the stored channel configuration since the channel uid isn't changing. I was able to test it with my setup.

https://github.com/openhab/openhab-addons/blob/d5e65dadea7331b92dbda82356b0fd0c7f79907a/bundles/org.openhab.binding.insteon/src/main/java/org/openhab/binding/insteon/internal/handler/InsteonDeviceHandler.java#L116-L125

https://github.com/openhab/openhab-addons/blob/d5e65dadea7331b92dbda82356b0fd0c7f79907a/bundles/org.openhab.binding.insteon/src/main/java/org/openhab/binding/insteon/internal/handler/InsteonLegacyDeviceHandler.java#L487-L498

Reply:
But the whole point of this migration is to keep the same channel id, to prevent any breakage, while migrating to the (new) legacy channel type. Again, I have fully tested myself and I had at least another user test it successfully as well.

So I don't understand how the scenario you outlined above applies here.

The legacy device migration works as follow:

1. The `InsteonDeviceHandler` identifies a given thing connected to a `network` legacy bridge and initiate the migration.

2. The channels configuration for the identified thing are stored in the `network` legacy bridge based on its current channel uid.

3. The thing handler `changeThingType` is initiated changing the thing type to `legacyDevice` with the current thing configuration. That (core) method changes the underlying thing type while keeping the current thing id.

4.  When the changed thing is initialized by `InsteonLegacyDeviceHandler`, it creates the channels by using the same channel uid (since same thing and channel ids) and the previous channel configuration retrieved from the `network` legacy bridge connecting with the (new) legacy channel types.

Reply:
When looking at the `changeThingType` i noticed the `legacy` prefix was added. Because the structure - by default - ties both the channel id and type id to the exact same name, i misread the implementation of `changeThingType`. I now see it only changes the type id.
Sorry for the confusion.

Reply:
> Because the structure - by default - ties both the channel id and type id to the exact same name

That used to be the case for both the legacy and new channel creation. For the former as you mentioned, the channel type id is determined by adding `legacy` as prefix of the channel id. For the latter, the channel id is converted from lower hyphen to camel case.